### PR TITLE
Deprecate add_guard

### PR DESCRIPTION
**Pull request by yamaguchi1024**

### CODE REVIEW

**Comment by yamaguchi1024** [comment]
@gilbo Can you take a brief look at the changes, before I change all the use cite of `add_guard` ?

**Review by alexreinking — DISMISSED** [review]

File: src/exo/API.py [hunk 1]
```diff
@@ -679,17 +679,19 @@ def fuse_if(self, if1, if2):
...
-    def add_loop(self, stmt, var, hi):
+    def add_loop(self, stmt, var, hi, guard=False):
         if not isinstance(stmt, str):
             raise TypeError("expected first arg to be a string")
         if not isinstance(var, str):
             raise TypeError("expected second arg to be a string")
         if not isinstance(hi, int):
             raise TypeError("currently, only constant bound is supported")
+        if not isinstance(guard, bool):
+            raise TypeError("guard needs to be True or False")
```
Comment (alexreinking) [inline src/exo/API.py:690]: Consider adding a test that tries setting `guard` to a non-bool.

**Comment by alexreinking** [comment]
Looks great!

**Comment by yamaguchi1024** [comment]
@alexreinking Thanks, I still need to remove more use cite of `add_guard` for this PR

**Comment by alexreinking** [comment]
I think it's fine to merge this first, but if you want to wait, that's fine, too.

**Comment by yamaguchi1024** [comment]
@gilbo @alexreinking I'll merge this PR if it looks alright~

File: tests/gemmini/conv/test_gemmini_conv.py [hunk 2]
```diff
@@ -1,448 +0,0 @@
-from __future__ import annotations
```
Comment (alexreinking) [inline tests/gemmini/conv/test_gemmini_conv.py:1]: Why was this file deleted?

File: tests/gemmini/conv/test_gemmini_conv_stride1.py [hunk 3]
```diff
@@ -262,124 +304,6 @@ def test_conv_3():
 
     print(conv)
 """
```
Comment (alexreinking) [inline tests/gemmini/conv/test_gemmini_conv_stride1.py:306]: Is this just an empty string now? If so, get rid of the string.

File: src/exo/API.py [hunk 4]
```diff
@@ -696,17 +672,19 @@ def fuse_if(self, if1, if2):
 
         return Procedure(loopir, _provenance_eq_Procedure=self)
 
-    def add_loop(self, stmt, var, hi):
+    def add_loop(self, stmt, var, hi, guard=False):
```
Comment (alexreinking) [inline src/exo/API.py:675]: At every call site, we're making sure to write `guard=True/False`. We should consider _requiring_ the use of the keyword by changing the signature as follows:

```suggestion
    def add_loop(self, stmt, var, hi, *, guard=False):
```
Arguments after the `*` are _keyword-only_.

**Review by alexreinking — CHANGES_REQUESTED** [review]
Looks pretty good, just a little clean-up and I'll approve.

File: tests/gemmini/conv/test_gemmini_conv.py (hunk 2 above)
Comment (yamaguchi1024) [reply tests/gemmini/conv/test_gemmini_conv.py:1]: This is a super outdated test that was skipped (see line 8). It's fine not to delete this file, but it doesn't matter

Comment (alexreinking) [reply tests/gemmini/conv/test_gemmini_conv.py:1]: Ok, I was just making sure it was intentional.

**Review by alexreinking — APPROVED** [review]
Looks good to me! Just did the merge with `master` myself, so we'll just wait for green. Go ahead and enable auto-merge.
